lmnet/lmnet/networks/classification/darknet.py

```diff
@@ -286,76 +287,39 @@ def __init__(
             activation_quantize_kwargs(dict): Initialize kwargs for activation quantizer.
         """
 
-        super().__init__(
+        Darknet.__init__(
+            self,
             *args,
             **kwargs,
         )
 
-        self.quantize_first_convolution = quantize_first_convolution
-        self.quantize_last_convolution = quantize_last_convolution
-
-        activation_quantizer_kwargs = activation_quantizer_kwargs if not None else {}
-        weight_quantizer_kwargs = weight_quantizer_kwargs if not None else {}
+        BaseQuantize.__init__(
```

Comment:
We need 2 initialization `Darknet.__init__()`  and `BaseQuantize.__init__()` ? .
Why don't use below
```
super().__init__(
           activation_quantizer,
           activation_quantizer_kwargs,
            weight_quantizer,
            weight_quantizer_kwargs,
            quantize_first_convolution,
            quantize_last_convolution,
            *args,
            **kwargs,
)
```
?

Reply:
Please see
https://www.ianlewis.org/en/mixins-and-python

Reply:
The reference tells that if you call a function appears in Mixins and base class (test function), only the function in the leftmost Mixins will be called, doesn't it?

Reply:
When someone create new network class and also want quantized one,
I hope to provide easy way to quantize, something like this.
```
class Newnet(Base):
    ...
    ...
    ...


class QuantizedNewnet(QuantizeMixin, Newnet):
   pass
```

Reply:
I'm not sure we can do this.I tried testing with a simple code below.
```
class base1:
    def __init__(self,i):
        print("base1")
        self.i = i

class base2:
    def __init__(self, j):
        print("base2")
        self.j = j

class child(base2, base1):
    pass
```
and found that when I initialize an object of child class, only `base2` is printed out.
which means that the init function of `base1` is not called by default and we have to call it manually.

Reply:
It seems like there is a way to do that.
I will let you know after testing it.

Reply:
When you use mix-in, I think second parent class should not have any instance variable.

For example, `class Child(P1, P2)`, `P2` class should only contain static methods.

In this PR,  `BaseQuantize` has only static method.
So, `__init__` method can be removed. This understand is wrong?

https://github.com/blue-oil/blueoil/blob/4b6f159bc619f3dae21041b1193c1b8bfdb0469b/lmnet/lmnet/networks/base_quantize.py#L20-L41

After remove `__init__` from `BaseQuantize`, next you can remove `BaseQuantize.__init__`.
Finally, you can use `super().__init__` in `DarknetQuantize`.

Reply:
> In this PR, `BaseQuantize` has only one static method.
> So, `__init__` method can be removed. This understand is wrong?
> 

The `BaseQuantize` contain one static method as you mentioned.
However, in my opinion, the __init__ method is still necessary because it is used to initialize variables for quantization, e.g. quantization function.

Reply:
Oh. Ok, I missed some code.

Do you talking about [base](https://github.com/blue-oil/blueoil/blob/4b6f159bc619f3dae21041b1193c1b8bfdb0469b/lmnet/lmnet/networks/classification/darknet.py#L314-L325) function using `self.weight_quantization`, `self.quantize_first_convolution`, `self.quantize_last_convolution`?

My new suggestion (psudecode) is here

1: Remove Multiple inheritance `DarknetQuantize(Darknet, BaseQuantize)` -> `DarknetQuantize(Darknet)`
2: Rename class name `BaseQuantize` -> `QuantizeParameters`
3: Add instance variables to `QuantizeParameters`
4: Add getter function to `QuantizeParameters`
5: 

```
self.quantize_params = QuantizeParameters(
    weight_quantization=self.weight_quantization,
    use_histogram=True,
    first_layer_name="block_1/",
    last_layer_name="conv_23/",
    quantize_first_convolution=self.quantize_first_convolution,
    quantize_last_convolution=self.quantize_last_convolution,
)
```

6: Simplify `base` function.

```
def base(self, images, is_training):
    with tf.variable_scope("", custom_getter=self.quantize_params.getter()): 
    return super().base(images, is_training)
```

This idea is little complicated.
Somebody may have more better idea...

Reply:
Thank you for your suggestion.
If I understand correctly, your code initializes quantization parameter in the `__init__` function of `DarknetQuantize` class, doesn't it?